### PR TITLE
Add controllers to map/redirect updated TIDs for legacy news feeds.

### PR DESCRIPTION
**Pull request by joeparsons**
In theory, this should make it possible for the legacy news feeds that various Quickstart sites rely on to continue to function after the launch of the new site even though the new site has different taxonomy term IDs than the old site.

This is an attempt to execute an idea @trackleft  described to me on Friday.

- Replaces `news_deprecated_feeds_terms` view with new controller that returns JSON response of terms data matching legacy feed that replaces TID values from the new site DB with legacy values (controller for `/feeds/json/categories/%`)
- Adds a new controller for `/feeds/json/stories/%` route that redirects requests to updated `news_deprecated_feeds_stories` view replacing the old TID value with the new site TID value.
- Updates the path for the `news_deprecated_feeds_stories` view to `feed/json/stories-updated/%`
- Adds two TID mapping files.  One uses old TIDs as keys and new TIDs as values and other uses new TIDs as keys and old TIDs as values.

### CODE REVIEW

File: src/Controller/LegacyStoriesRedirectController.php [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+    else {
+      $module_path = $this->moduleHandler->getModule('news_legacy_feeds')->getPath();
+      $mapping_file_path = "$module_path/mapping_data/new_tid_mappings.json";
+
+      $new_tid_mappings = json_decode(file_get_contents($mapping_file_path), TRUE);
+      $this->cache->set('news_legacy_feeds.new_tid_mappings', $new_tid_mappings);
+    }
+
+    $new_tid = isset($new_tid_mappings[$tid]) ? $new_tid_mappings[$tid] : $tid;
+    $new_url = Url::fromRoute('view.news_deprecated_feeds_stories.category', ['term_node_tid_depth' => $new_tid])->toString();
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:82]:
```suggestion
    $new_url = Url::fromUri('/feed/json/stories-updated/' . $new_tid);
```

File: news_legacy_feeds.routing.yml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+news_legacy_feeds.legacy_categories:
+  path: '/feed/json/categories/{vocabularies}'
```
Comment (joeparsons) [inline news_legacy_feeds.routing.yml:2]:
```suggestion
  path: '/feed/json/categories/{vocabulariesParam}'
```

File: src/Controller/LegacyCategoriesController.php [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+      $oldTidMappings = json_decode($json, TRUE);
+      $this->cacheBackend->set('news_legacy_feeds.old_tid_mappings', $oldTidMappings);
+    }
+
+    $data = [];
+    foreach ($vocabularies as $vocabulary) {
+      $terms = $this->entityTypeManager->getStorage('taxonomy_term')->loadTree($vocabulary, 0, NULL, TRUE);
+
+      foreach ($terms as $term) {
+        $tid = $term->id();
```
Comment (joeparsons) [inline src/Controller/LegacyCategoriesController.php:106]:
```suggestion
        $tid = $new_tid = $term->id();
```

File: src/Controller/LegacyCategoriesController.php [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        // Replace TID if it's in the old mappings.
+        if (isset($oldTidMappings[$tid])) {
+          $tid = $oldTidMappings[$tid];
+        }
+
+        $url = Url::fromRoute('entity.taxonomy_term.canonical', ['taxonomy_term' => $tid]);
```
Comment (joeparsons) [inline src/Controller/LegacyCategoriesController.php:112]:
```suggestion
        $url = Url::fromRoute('entity.taxonomy_term.canonical', ['taxonomy_term' => $new_tid]);
```

File: src/Controller/LegacyStoriesRedirectController.php [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+
+  /**
+   * Redirects to the new story page based on the old term IDs.
+   *
+   * @param string $termIdsParam
+   *   The old term IDs separated by " " ("+" in the raw request).
+   * @return \Symfony\Component\HttpFoundation\RedirectResponse
+   *   A redirect response object.
+   */
+  public function redirectStory($termIdsParam) {
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:64]:
```suggestion
  public function redirectTids($termIdsParam) {
```

File: src/Controller/LegacyStoriesRedirectController.php [hunk 6]
```diff
@@ -0,0 +1,91 @@
...
+use Drupal\Core\Controller\ControllerBase;
+use Drupal\Core\Extension\ModuleHandlerInterface;
+use Drupal\Core\Url;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+use Symfony\Component\HttpFoundation\RedirectResponse;
+
+/**
+ * Class LegacyStoriesRedirectController.
+ *
+ * Handles redirection for legacy stories.
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:15]:
```suggestion
 * Handles redirection for legacy stories feed based on term ID mappings
```

File: src/Controller/LegacyStoriesRedirectController.php [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+   */
+  public static function create(ContainerInterface $container) {
+    return new static(
+      $container->get('cache.default'),
+      $container->get('module_handler')
+    );
+  }
+
+  /**
+   * Redirects to the new story page based on the old term IDs.
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:57]:
```suggestion
   * Redirects the request based on term ID mappings.
```

File: src/Controller/LegacyStoriesRedirectController.php [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+      $this->cacheBackend->set('news_legacy_feeds.new_tid_mappings', $newTidMappings);
+    }
+
+    $newTermIds = [];
+    foreach ($termIds as $termId) {
+      // Replace TID if it's in the new mappings, otherwise preserve it as-is.
+      $newTermIds[] = isset($newTidMappings[$termId]) ? $newTidMappings[$termId] : $termId;
+    }
+
+    // Redirect to the new story page.
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:86]:
```suggestion
    // Redirect to the new stories feed view with updated TIDs.
```

**Review by joeparsons — COMMENTED** [review]
Restore various comments.

File: src/Controller/LegacyCategoriesController.php [hunk 9]
```diff
@@ -0,0 +1,132 @@
...
+      'uanews_tags' => 'az_news_tags',
+      'uanews_categories' => 'news_story_categories',
+      'event_categories' => 'az_event_categories',
+    ];
+
+    if ($vocabulariesParam === 'all') {
+      $vocabularies = array_keys($supportedVocabularies);
+    }
+    else {
+      $vocabularies = explode(' ', $vocabulariesParam);
```
Comment (joeparsons) [inline src/Controller/LegacyCategoriesController.php:87]:
```suggestion
      // Plus characters in request parameters turn into spaces for some reason.
      $vocabularies = explode(' ', $vocabulariesParam);
```

File: news_legacy_feeds.routing.yml [hunk 10]
```diff
@@ -0,0 +1,14 @@
+news_legacy_feeds.legacy_categories:
+  path: '/feed/json/categories/{vocabulariesParam}'
+  defaults:
+    _controller: '\Drupal\news_legacy_feeds\Controller\LegacyCategoriesController::getTaxonomyTerms'
+  requirements:
+    _permission: 'access content'
+
+news_legacy_feeds.legacy_stories_redirect:
+  path: '/feed/json/stories/{termIdsParam}'
```
Comment (joeparsons) [inline news_legacy_feeds.routing.yml:9]:
```suggestion
  path: '/feed/json/stories/id/{termIdsParam}'
```

File: src/Controller/LegacyStoriesRedirectController.php [hunk 11]
```diff
@@ -0,0 +1,91 @@
...
+    }
+
+    $newTermIds = [];
+    foreach ($termIds as $termId) {
+      // Replace TID if it's in the new mappings, otherwise preserve it as-is.
+      $newTermIds[] = isset($newTidMappings[$termId]) ? $newTidMappings[$termId] : $termId;
+    }
+
+    // Redirect to the new stories feed view with updated TIDs.
+    $redirectUrl = Url::fromUri('internal:/feed/json/stories-updated/' . implode('+', $newTermIds));
```
Comment (joeparsons) [inline src/Controller/LegacyStoriesRedirectController.php:87]:
```suggestion
    $redirectUrl = Url::fromUri('internal:/feed/json/stories/id-updated/' . implode('+', $newTermIds));
```

File: config/install/views.view.news_deprecated_feeds_stories.yml [hunk 12]
```diff
@@ -438,3 +438,99 @@ display:
         - user.permissions
       tags:
         - 'config:field.storage.node.field_az_photos'
+  rest_export_1:
+    id: rest_export_1
```
Comment (joeparsons) [inline config/install/views.view.news_deprecated_feeds_stories.yml:442]:
```suggestion
  website:
    id: website
```

File: src/EventSubscriber/ResponseWrapperSubscriber.php [hunk 13]
```diff
@@ -22,17 +22,6 @@ public function onResponse(ResponseEvent $event) {
...
-    $response = $event->getResponse();
-
-    if ($response->headers->get('Content-Type') === 'application/json') {
-      $data = json_decode($response->getContent(), true);
-      $wrappedData = ['terms' => $data];
-      $response->setContent(json_encode($wrappedData));
-      $event->setResponse($response);
-    }
-  }
   if ($route_name === 'view.news_deprecated_feeds_stories.category') {
```
Comment (joeparsons) [inline src/EventSubscriber/ResponseWrapperSubscriber.php:25]:
```suggestion
  if ($route_name === 'view.news_deprecated_feeds_stories.category' || 'view.news_deprecated_feeds_stories.category.website') {
```

File: src/EventSubscriber/ResponseWrapperSubscriber.php [hunk 14]
```diff
@@ -22,18 +22,10 @@ public function onResponse(ResponseEvent $event) {
...
-      $data = json_decode($response->getContent(), true);
-      $wrappedData = ['terms' => $data];
-      $response->setContent(json_encode($wrappedData));
-      $event->setResponse($response);
-    }
-  }
-  if ($route_name === 'view.news_deprecated_feeds_stories.category') {
+  if (
+    $route_name === 'view.news_deprecated_feeds_stories.category' ||
+    $route_name === 'view.news_deprecated_feeds_stories.category.website'
```
Comment (joeparsons) [inline src/EventSubscriber/ResponseWrapperSubscriber.php:27]:
```suggestion
    $route_name === 'view.news_deprecated_feeds_stories.website'
```

**Review by trackleft — APPROVED** [review]

**Pull request merged**